### ISSUE TITLE
Not receiving e-mails

### ISSUE THREAD

**Issue by m-rm** [dak180/FreeNAS-Report#40]
The script was working fine a few days ago, now suddenly I don't receive e-mails anymore. Things that may have something to do with it:

- Updated TrueNAS scale to the latest version
- Change SMTP credentials
- Moved script and config to a new location

Behavior:

- I receive test e-mails and alerts from TrueNAS just fine, it's only the script that's affected
- The script runs, no error is generated
- Log files in the /tmp folder are generated
- When I use sendmail with the generated log file I do not receive an e-mail
- When I use sendmail with a simple test file I do receive the e-mail

Script version: 1.8(.1)
TrueNAS Scale version; ElectricEel-24.10.1

**Comment by m-rm** [comment]
Log file: [20241222231828_logfilename.txt](https://github.com/user-attachments/files/18223408/20241222231828_logfilename.txt)

(I replaced my mail address for privacy reasons.)

**Comment by dak180** [comment]
> The script was working fine a few days ago, now suddenly I don't receive e-mails anymore. Things that may have something to do with it:
> 
> * Updated TrueNAS scale to the latest version

[See](https://forums.truenas.com/t/memes-truenas-zfs-and-related-share-your-own/335/302?u=dak180) [several](https://forums.truenas.com/t/multi-report/1302/225?u=dak180) posts on this (`sendmail` [is gone and not coming back](https://ixsystems.atlassian.net/browse/NAS-133174)). It is a known issue but I have a very limited ability to test fixes because I am running core and am unlikely to switch before the next major version of scale (and am most likely to switch only when I can no longer run the jails I want to). That said how willing are you to do testing/investigating in service of a fix? A good start would be to attach a copy of the man page of `exim` on the latest version of scale.

**Comment by JoeSchmuck** [comment]
We are working is, there are many people interested in how to send an email correctly in TrueNAS 24.10.1 and later.  Once we figure it out, we will share the love.  I am hoping for a direct replacement for sendmail, but I may be wishful thinking.  No one wants to redo their code. For now the best option is to roll back to 24.10.0.2 and give use a few weeks to sort this out.  read the threads @dak180 listed.  It is new security change and it will take time.

**Comment by m-rm** [comment]
> [See](https://forums.truenas.com/t/memes-truenas-zfs-and-related-share-your-own/335/302?u=dak180) [several](https://forums.truenas.com/t/multi-report/1302/225?u=dak180) posts on this (`sendmail` [is gone and not coming back](https://ixsystems.atlassian.net/browse/NAS-133174)). It is a known issue but I have a very limited ability to test fixes because I am running core and am unlikely to switch before the next major version of scale (and am most likely to switch only when I can no longer run the jails I want to). That said how willing are you to do testing/investigating in service of a fix? A good start would be to attach a copy of the man page of `exim` on the latest version of scale.

Oh, I didn't think it was a widespread problem. Here is a copy of the man page: [man_exim.txt](https://github.com/user-attachments/files/18234128/man_exim.txt)

>  I am hoping for a direct replacement for sendmail, but I may be wishful thinking.

What I don't understand is that sendmail works in some cases (at least for me). If I copy this into a file:

```
Subject: Terminal Email Send

Email Content line 1
Email Content line 2
```

and send it with `sudo sendmail not-my-email-address@gmail.com < email.txt`, I receive the mail.

**Comment by FDrebin** [comment]
Keeping an eye on this mess, just realized my emails weren't sending and went to the internet and here we are.

**Comment by dak180** [comment]
Waiting on oxyde1989/standalone-tn-send-email#3.

**Comment by dak180** [comment]
Any / every one on scale could you please post the results of `cat /etc/version`?

**Comment by krin-san** [comment]
```
$ cat /etc/version
24.10.1%
```
`%` because there's EOF on the first line.

**Comment by m-rm** [comment]
Same for me:
```
truenas% cat /etc/version
24.10.1%
```

**Comment by JoeSchmuck** [comment]
Use cat /etc/version | cut -d " " -f1 | sed 's/TrueNAS-//'
Hopefully that does it.  You may not need the | sed......

**Comment by FDrebin** [comment]
Performing `cat /etc/version | cut -d " " -f1 | sed 's/TrueNAS-//'` yields a result of `24.10.1`

**Comment by dak180** [comment]
If anyone would like to try out afed43e6d8 and tell me how/if it works I would appreciate it since I cannot .

**Comment by FDrebin** [comment]
> If anyone would like to try out [afed43e](https://github.com/dak180/FreeNAS-Report/commit/afed43e6d8d0c0121cbab640353e6c7cf53ea19d) and tell me how/if it works I would appreciate it since I cannot .

So I think either my conf needs to be updated to something else, or something is still off

`usage: multireport_sendemail.py [-h] [--subject SUBJECT] [--to_address TO_ADDRESS] [--mail_body_html MAIL_BODY_HTML]
                                [--attachment_files [ATTACHMENT_FILES ...]] [--mail_bulk MAIL_BULK]
multireport_sendemail.py: error: unrecognized arguments: --bulk_email`

**Comment by dak180** [comment]
> or something is still off

You need the latest version of [multireport_sendemail.py](https://github.com/oxyde1989/standalone-tn-send-email/releases)

**Comment by FDrebin** [comment]
> > or something is still off
> 
> You need the latest version of [multireport_sendemail.py](https://github.com/oxyde1989/standalone-tn-send-email/releases)

This is the one I just downloaded, I haven't messed with any of the beta fixes. This is the first time I've updated my TrueNAS Scale script(s) since this happened.

**Comment by FDrebin** [comment]
I just downloaded the latest files the above for report.sh and the latest sendmail.py , it did complain at first of where to place the sendmail.py and I did correct it, but same error.

**Comment by dak180** [comment]
> I just downloaded the latest files the above for report.sh and the latest sendmail.py , it did complain at first of where to place the sendmail.py and I did correct it, but same error.

Try 7748111.

**Comment by FDrebin** [comment]
> > I just downloaded the latest files the above for report.sh and the latest sendmail.py , it did complain at first of where to place the sendmail.py and I did correct it, but same error.
> 
> Try [7748111](https://github.com/dak180/FreeNAS-Report/commit/77481110c01d73c01c3c137cf88f03296400042b).

Thanks for the quick update! This one resulted in the following : 

`sudo ./report.sh -c /mnt/Tank1/Home/secretuser/scripts/report.cfg
sudo: process 584494 unexpected status 0x57f
./report.sh: line 2800: 584494 Killed                  python3 multireport_sendemail.py --bulk_email "$(base64 -w 0 < "${logfile}")"`

**Comment by FDrebin** [comment]
So with commit https://github.com/dak180/FreeNAS-Report/commit/37fd579c7a68349496132fbaaadd0ef9202a667d it no longer generates an error, but the log doesnt appear to contain anything useful, but no email is sent.

**Comment by dak180** [comment]
37fd579 is not part of the [topic/sendmail](https://github.com/dak180/FreeNAS-Report/tree/topic/sendmail) dev branch for this.

**Comment by FDrebin** [comment]
Ah ok understood. With the latest changes made to the sendmail.py script, I was able to send the email but only using the multi_report script. I've been unable to get the report.sh that you maintain dak to send the mail successfully. at least so far.

**Comment by m-rm** [comment]
> > I just downloaded the latest files the above for report.sh and the latest sendmail.py , it did complain at first of where to place the sendmail.py and I did correct it, but same error.
> 
> Try [7748111](https://github.com/dak180/FreeNAS-Report/commit/77481110c01d73c01c3c137cf88f03296400042b).

Its printing a lot of encoded stuff and then this:
```
AgUkVBRCBETUEKClRlc3RfRGVzY3JpcHRpb24gICAgU3RhdHVzICAgICAgICAgICAgICAgICAgUmVtYWluaW5nICBMaWZlVGltZShob3VycykgIExCQV9vZl9maXJzdF9lcnJvcgpTaG9ydCBvZmZsaW5lICAgICAgIENvbXBsZXRlZCB3aXRob3V0IGVycm9yICAgICAgIDAwJSAgICAgICAyNTQgICAgICAgICAtCjxicj48YnI+CjwvcHJlPgotLThiYTFkYjFmYWY0MzVkYWU1NGI2MDE0YzY3OGMwZTI4LS0K"
./report.sh: line 2800: 1469633 Killed                  python3 multireport_sendemail.py --bulk_email "$(base64 -w 0 < "${logfile}")"
```

No e-mail received.

**Comment by dak180** [comment]
0864f1b is up for testing.

**Comment by m-rm** [comment]
> [0864f1b](https://github.com/dak180/FreeNAS-Report/commit/0864f1b92768d1e0f5e88653f89dec86f362263e) is up for testing.

First try:
```
python3: can't open file '/mnt/SSD/Private/home/madmin/scripts/multireport_sendemail.py': [Errno 2] No such file or directory
```

After moving multireport_sendemail.py to the script folder instead of ./usr/bin:
```
usage: multireport_sendemail.py [-h] [--subject SUBJECT] [--to_address TO_ADDRESS] [--mail_body_html MAIL_BODY_HTML]
                                [--attachment_files [ATTACHMENT_FILES ...]] [--mail_bulk MAIL_BULK]
multireport_sendemail.py: error: unrecognized arguments: --bulk_email /tmp/20250119115415_logfilename.tmp.64
```

After changing the parameter from --bulk_email to --mail_bulk:
```
{"error": false, "detail": ">> All is Good <<", "logfile": "/mnt/SSD/Private/home/madmin/scripts/sendemail_log/20250119_115847.txt", "total_attach": 0, "ok_attach": 0}
```

And I received the e-mail! Nice work!

**Comment by dak180** [comment]
The way this works now is that `multireport_sendemail.py` needs to be in your path (a location will be suggested if it is not) and be at least version 0.08 or higher.

**Comment by m-rm** [comment]
https://github.com/dak180/FreeNAS-Report/commit/b8a80a0945a269604adf7a3d296b1e287340ad6f is not working for me. I updated multireport_sendemail.py to 0.08.

```
sudo: process 3073592 unexpected status 0x57f
./report.sh: line 2805: 3073592 Killed                  multireport_sendemail.py --mail_bulk "${logfile}.64"
```

**Comment by dak180** [comment]
> After moving multireport_sendemail.py to the script folder instead of ./usr/bin:

@m-rm did you update the one in the scripts folder (which is not used) or the one in the path (`./usr/bin`)?

**Comment by FDrebin** [comment]
Running the latest avail report.sh with a properly placed multireport_sendmail.py I receive : 

`sudo ./report.sh -c /mnt/Tank1/Houses/secretuser/scripts/report.cfg
{"error": true, "detail": "KO: cannot access local variable 'hostname' where it is not associated with a value", "logfile": "/mnt/Tank1/Houses/secretuser/scripts/sendemail_log/20250119_134913.txt", "total_attach": 0, "ok_attach": 0}`

**Comment by m-rm** [comment]
> > After moving multireport_sendemail.py to the script folder instead of ./usr/bin:
> 
> [@m-rm](https://github.com/m-rm) did you update the one in the scripts folder (which is not used) or the one in the path (`./usr/bin`)?

Oh you are absolutely right, that was the problem. Running from console now works fine. But when running via cron I get:

```
Traceback (most recent call last):
  File "/mnt/SSD/Private/home/madmin/scripts/usr/bin/multireport_sendemail.py", line 333, in <module>
    log_file, log_file_count = create_log_file()
                               ^^^^^^^^^^^^^^^^^
  File "/mnt/SSD/Private/home/madmin/scripts/usr/bin/multireport_sendemail.py", line 34, in create_log_file
    os.makedirs(log_dir)
  File "<frozen os>", line 225, in makedirs
OSError: [Errno 30] Read-only file system: '/sendemail_log'

During handling of the above exception, another exception occurred:
... 16 more lines ...

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/mnt/SSD/Private/home/madmin/scripts/usr/bin/multireport_sendemail.py", line 361, in <module>
    process_output(True, f"Error: {e}", 1)
  File "/mnt/SSD/Private/home/madmin/scripts/usr/bin/multireport_sendemail.py", line 67, in process_output
    response = json.dumps({"error": error, "detail": detail, "logfile": log_file, "total_attach": attachment_count, "ok_attach": attachment_ok_count}, ensure_ascii=False)
                                                                        ^^^^^^^^
NameError: name 'log_file' is not defined
```

I guess the working directory is / when running in cron and it tries to create /sendemail_log which fails?

**Comment by dak180** [comment]
> I guess the working directory is / when running in cron and it tries to create /sendemail_log which fails?

@m-rm  Try 1ba2328.

@FDrebin what is the version of `multireport_sendmail.py` you are using (does the file start with `#!`)?

**Comment by m-rm** [comment]
> > I guess the working directory is / when running in cron and it tries to create /sendemail_log which fails?
> 
> [@m-rm](https://github.com/m-rm) Try [1ba2328](https://github.com/dak180/FreeNAS-Report/commit/1ba23287db2ab8fd716712b255be29a2ac69623e).

It's working! Thank you very much!

**Comment by FDrebin** [comment]
> > I guess the working directory is / when running in cron and it tries to create /sendemail_log which fails?
> 
> [@m-rm](https://github.com/m-rm) Try [1ba2328](https://github.com/dak180/FreeNAS-Report/commit/1ba23287db2ab8fd716712b255be29a2ac69623e).
> 
> [@FDrebin](https://github.com/FDrebin) what is the version of `multireport_sendmail.py` you are using (does the file start with `#!`)?

The .py script does have the #!/usr/bin/env python3 and it is version V 0.09

**Comment by dak180** [comment]
> The .py script does have the #!/usr/bin/env python3 and it is version V 0.09

Try version 0.08; if that works then it may be a `multireport_sendmail.py` [issue](https://github.com/oxyde1989/standalone-tn-send-email/issues/new/choose).

**Comment by mkninc** [comment]
> Running the latest avail report.sh with a properly placed multireport_sendmail.py I receive :
> 
> `sudo ./report.sh -c /mnt/Tank1/Houses/secretuser/scripts/report.cfg {"error": true, "detail": "KO: cannot access local variable 'hostname' where it is not associated with a value", "logfile": "/mnt/Tank1/Houses/secretuser/scripts/sendemail_log/20250119_134913.txt", "total_attach": 0, "ok_attach": 0}`

@FDrebin Where you able to resolve this issue?
I currently have the same problem.

**Comment by FDrebin** [comment]
> > Running the latest avail report.sh with a properly placed multireport_sendmail.py I receive :
> > `sudo ./report.sh -c /mnt/Tank1/Houses/secretuser/scripts/report.cfg {"error": true, "detail": "KO: cannot access local variable 'hostname' where it is not associated with a value", "logfile": "/mnt/Tank1/Houses/secretuser/scripts/sendemail_log/20250119_134913.txt", "total_attach": 0, "ok_attach": 0}`
> 
> [@FDrebin](https://github.com/FDrebin) Where you able to resolve this issue? I currently have the same problem.

I did not resolve it, I figured I would just use the Multi Report for now.

**Comment by dak180** [comment]
> I did not resolve it, I figured I would just use the Multi Report for now.

@FDrebin did you try v0.08 of `multireport_sendmail.py`?

**Comment by FDrebin** [comment]
> > I did not resolve it, I figured I would just use the Multi Report for now.
> 
> [@FDrebin](https://github.com/FDrebin) did you try v0.08 of `multireport_sendmail.py`?

Yes, but no email was sent. I will give it a try again today once I get all my ducks in a row.

**Comment by FDrebin** [comment]
> > I did not resolve it, I figured I would just use the Multi Report for now.
> 
> [@FDrebin](https://github.com/FDrebin) did you try v0.08 of `multireport_sendmail.py`?

No luck with version 0.08 of the multireport_sendmail.py , the multireport script does still work as expected.

**Comment by FDrebin** [comment]
I still receive : 
`$ sudo ./report.sh -c report.cfg
{"error": true, "detail": "KO: cannot access local variable 'hostname' where it is not associated with a value", "logfile": "/mnt/BigData/Houses/runner/scripts/sendemail_log/20250124_150737.txt", "total_attach": 0, "ok_attach": 0}`

The log doesn't contain much more useful data either : 
`File /mnt/BigData/Houses/runner/scripts/sendemail_log/20250124_150737.txt successfully generated
10 totals file log
0 totals attachment
trying read mail.config
read mail.config successfully
Switching for the right provider
** SMTP Version **
parsing smtp config
switch from classic send and bulk email
using bulk email provided
body is a file
Base64 message is valid.
establing connection based on security level set on TN: TLS
entered TLS path
adding ehlo to the message
KO: cannot access local variable 'hostname' where it is not associated with a value`

**Comment by m-rm** [comment]
I think I found the problem. For me it works because I use GMail OAuth to send the e-mail. When SMTP is used together with the `bulk_email` option there is a bug in `multireport_sendmail.py`. I created an issue https://github.com/oxyde1989/standalone-tn-send-email/issues/6.

**Comment by FDrebin** [comment]
> I think I found the problem. For me it works because I use GMail OAuth to send the e-mail. When SMTP is used together with the `bulk_email` option there is a bug in `multireport_sendmail.py`. I created an issue [oxyde1989/standalone-tn-send-email#6](https://github.com/oxyde1989/standalone-tn-send-email/issues/6).

Interesting! Because that was looked at on the TrueNAS discussion page, and I had the same issue as other individuals as I use SMTP and it works with the multi report, so I wonder why it doesn't work here.

**Comment by m-rm** [comment]
I haven't looked at the multi report source code but I guess it probably doesn't use the bulk_email parameter.

**Comment by FDrebin** [comment]
Ok so, the dev for the py script did an update, I have the update and re-ran report.sh and I receive the following message now : 

`sudo ./report.sh -c report.cfg
{"error": true, "detail": "KO: 'str' object has no attribute 'as_string'", "logfile": "/mnt/BigData/Houses/runner/scripts/sendemail_log/20250125_095405.txt", "total_attach": 0, "ok_attach": 0}`

**Comment by dak180** [comment]
@FDrebin any error that you get in the form of json is being reported by `multireport_sendmail.py` and should be reported there: oxyde1989/standalone-tn-send-email#6.

**Comment by m-rm** [comment]
> Ok so, the dev for the py script did an update, I have the update and re-ran report.sh and I receive the following message now :
> 
> `sudo ./report.sh -c report.cfg {"error": true, "detail": "KO: 'str' object has no attribute 'as_string'", "logfile": "/mnt/BigData/Houses/runner/scripts/sendemail_log/20250125_095405.txt", "total_attach": 0, "ok_attach": 0}`

@FDrebin v0.12 of multireport_sendmail.py was just released a few minutes ago that seems to fix this exact issue.

**Comment by FDrebin** [comment]
Success! Yes, things are working again with the report.sh and the existing report.cfg + the latest v0.12 py script.

**Comment by krin-san** [comment]
```
sudo: process 2456779 unexpected status 0x57f
```
This error happens when a locally cloned `standalone-tn-send-email/multireport_sendemail.py` (v0.13) doesn't have X (execute) permission.

Here's how I found it out and fixed it:
```
# /root/bin/report.sh -c /root/bin/report.cfg
sudo: process 2456779 unexpected status 0x57f
/root/bin/report.sh: line 2812: 2456779 Killed                  ( cd "${logfileLocation}"; base64 -w 0 < "${logfile}" > "${logfile}.64"; multireport_sendemail.py -
-mail_bulk "${logfile}.64" )

# cd /tmp

# multireport_sendemail.py --mail_bulk 20250126165411_report.log.tmp.64
sudo: process 2477210 unexpected status 0x57f
zsh: killed     multireport_sendemail.py --mail_bulk 20250126165411_report.log.tmp.64

# ls -l /root/bin/
...
lrwxrwxrwx 1 root root      49 Jan 19 12:00 multireport_sendemail.py -> standalone-tn-send-email/multireport_sendemail.py

# ls -l /root/bin/standalone-tn-send-email 
total 68
-rw-r--r-- 1 root root 20589 Jan 26 16:48 multireport_sendemail.py

# chmod ug+x /root/bin/standalone-tn-send-email/multireport_sendemail.py 

# multireport_sendemail.py --mail_bulk 20250126165411_report.log.tmp.64               

Security Warning: this folder is accessible to non-priviliged users that are nor owner or in group
{"error": false, "detail": "<< Email Sent >>\nSecurity Warning: this folder is accessible to non-priviliged users that are nor owner or in group", "logfile": "/tmp/sendemail_log/20250126_165929.txt", "total_attach": 0, "ok_attach": 0}
root@nas[/tmp]# 
```

**Comment by dak180** [comment]
> This error happens when a locally cloned `standalone-tn-send-email/multireport_sendemail.py` (v0.13) doesn't have X (execute) permission.

The script [already tries to do this](https://github.com/dak180/FreeNAS-Report/blob/master/report.sh#L2464) for you. This depends on you having the script directly in your path (such as the path it suggests). You could also suggest to oxyde1989 that the executable bit be set in the git repo.

**Comment by krin-san** [comment]
> > This error happens when a locally cloned `standalone-tn-send-email/multireport_sendemail.py` (v0.13) doesn't have X (execute) permission.
> 
> The script [already tries to do this](https://github.com/dak180/FreeNAS-Report/blob/master/report.sh#L2464) for you. This depends on you having the script directly in your path (such as the path it suggests). You could also suggest to oxyde1989 that the executable bit be set in the git repo.

That workaround didn't work because symlink is executable:
```
lrwxrwxrwx 1 root root      49 Jan 19 12:00 multireport_sendemail.py -> standalone-tn-send-email/multireport_sendemail.py
```

Thanks for the "suggestion" idea – I have opened a PR https://github.com/oxyde1989/standalone-tn-send-email/pull/7 to make my suggestion just one click away from the reality.
